src/wpool/mongoose_gun_worker.erl

```diff
@@ -0,0 +1,276 @@
+    end;
+
+%% `gun_up' informs the owner process that the connection or reconnection completed.
+handle_info({gun_up, ConnPid, Protocol},
+            State = #gun_worker_state{pid = ConnPid}) ->
+    ?DEBUG("gun_up in mongoose_gun_worker. Connection is up with PID: ~p and protocol ~p", [ConnPid, Protocol]),
+    {noreply, State};
+
+%% `gun_down' is a message informing that Gun has lost the connection.
+%% It may try to reconnect, according to the `retry' and `retry_timeout' options,
+%% passed in the init options.
+handle_info({gun_down, PID, Protocol, Reason, KilledStreams, UnprocessedStreams}, State) ->
+    ?WARNING_MSG("gun_down in mongoose_gun_worker. "
+                 "Gun has lost the ~p connection ~p because of \"~p\", killing ~p streams.",
+                 [Protocol, PID, Reason, length(KilledStreams) + length(UnprocessedStreams)]),
+    ?DEBUG("gun_down in mongoose_gun_worker. Killed streams: ~p. Unprocessed streams: ~p",
+           [KilledStreams, UnprocessedStreams]),
+    {noreply, State};
+
+%% `gun_error' is a message informing about any errors concerning connections or
+%% streams handled by Gun.
+handle_info({gun_error, ConnPid, Reason},
+            State = #gun_worker_state{pid = ConnPid}) ->
+    ?WARNING_MSG("gun_error in mongoose_gun_worker. Reason: ~p.", [Reason]),
+    {noreply, State};
+handle_info({gun_error, ConnPid, StreamRef, Reason},
+            State = #gun_worker_state{pid = ConnPid}) ->
+    ?WARNING_MSG("gun_error in mongoose_gun_worker. Reason: ~p. Stream reference: ~p",
+                 [Reason, StreamRef]),
+    {noreply, State};
+
+%% After `retry' number of `gun_down' messages, the connection process dies.
+%% Because it is monitored, we receive a `DOWN' message and may restart the
+%% connection.
+handle_info({'DOWN', MRef, process, ConnPid, Reason},
+            #gun_worker_state{pid = ConnPid, monitor = MRef} = State) ->
+    ?WARNING_MSG("Mongoose_gun_worker has lost the connection with PID ~p. Reason ~p.",
+                 [ConnPid, Reason]),
+    ConnectedState = launch_connection(State),
+    NewState = retry_all(ConnectedState),
+    {noreply, NewState};
+
+handle_info(M, S) ->
+    ?ERROR_MSG("Unexpected message in gun_worker ~p", [M]),
+    {noreply, S}.
+
+terminate(_Reason, #gun_worker_state{pid = undefined}) ->
+    ok;
+terminate(_Reason, #gun_worker_state{pid = GunPid, requests = Requests}) ->
+    maps:fold(
+     fun(StreamRef, {_Request, #response_data{from = From, timeout_timer = TRef}}, _) ->
+             erlang:cancel_timer(TRef),
+             gun:cancel(GunPid, StreamRef),
+             gun:flush(StreamRef),
+             gen_server:reply(From, {error, request_timeout})
+     end, ok, Requests),
+    gun:flush(GunPid),
+    gun:shutdown(GunPid).
+
+%%%===================================================================
+%%% Internal functions
+%%%===================================================================
+
+parse_uri(Host) ->
```

Comment:
Maybe the name of the function could be changed as it also reads the default opts.